### PR TITLE
Modernize

### LINKED ISSUE #9: Upgrade to `redis@3`
There's a vulnerability alert against `redis@2` that's being used now: https://github.com/digitalbazaar/bedrock-redis/security/dependabot/package.json/redis/open

https://github.com/NodeRedis/node-redis/blob/master/CHANGELOG.md#v300---09-feb-2020

I think we would just bump this dep and release it as a breaking 4.x release here.  Breaking since several deprecated client APIs have been removed and we expose the client directly via this module.

The promises wrapper may need attention as well:
https://github.com/digitalbazaar/bedrock-redis/blob/master/package.json#L27

### PR DESCRIPTION
**Pull request by tolson4**
This pr does a few things to bring this package up to date with the rest of the bedrock ecosystem.

- Update to ESM
- Switches to native `node-redis` [client](https://github.com/redis/node-redis/blob/master/docs/v3-to-v4.md) since it is promise based now, drops `promise-redis`  dependency
- Pull in latest bedrock deps
- Update [backoff](https://github.com/tim-kos/node-retry) package and associated backoff/reconnect logic and configuration

indirectly resolves #9

### CODE REVIEW

**Comment by tolson4** [comment]
Results of tests are available on our fork: https://github.com/dogwoodlogic/bedrock-redis/actions/runs/3158929020

File: lib/client.js [hunk 1]
```diff
@@ -1,68 +1,37 @@
...
+import logger from './logger.js';
 
-const {config} = require('bedrock');
-const logger = require('./logger');
-const redis = require('promise-redis')();
-const Backoff = require('backo');
-
-module.exports = class Client {
+import {createClient} from 'redis';
+export class Client {
```
Comment (dlongley) [inline lib/client.js]:
```suggestion
import {createClient} from 'redis';

export class Client {
```

File: lib/client.js [hunk 2]
```diff
@@ -1,68 +1,37 @@
...
-          logger.error('exceeded max connection times', {
-            totalRetryTime: options.total_retry_time,
-            timesConnected: options.times_connected
-          });
-          // FIXME: improve error handling
-          return new Error('redis: exceed max connection times');
+    logger.debug(`initializing client at: ${host}:${port}`);
+    this.client = createClient({
+      socket: {host, port, reconnectStrategy: attempt => {
+        if(attempt > config.redis.retryOptions.retries) {
```
Comment (dlongley) [inline lib/client.js]: Can't suggest for the whole block here, but here's a start to bring it inline with our style:

```suggestion
      socket: {
        host,
        port,
        reconnectStrategy(attempt) {
          if(attempt > config.redis.retryOptions.retries) {
```

File: lib/client.js [hunk 3]
```diff
@@ -1,68 +1,37 @@
...
 
-    // FIXME: improve error handling
-    this.client.on('error', error => logger.error(`redis error: ${error}`));
-    this.client.on('end', () => logger.verbose('connection closed'));
-    this.client.on('warning', msg => logger.warning(`redis warning: ${msg}`));
+  async connect() {
+    try {
+      await this.client.connect();
+    } catch(error) {
+      logger.error('Unable to connect to redis.', {error});
```
Comment (dlongley) [inline lib/client.js]: This seems like a problem -- if there's an error when connecting it is never surfaced outside of this call. This should rethrow the error:

```suggestion
      logger.error('Unable to connect to redis.', {error});
      throw error;
```

File: lib/config.js [hunk 4]
```diff
@@ -1,33 +1,21 @@
...
-config.redis.retry = {
+config.redis.retryOptions = {
   // min retry delay (ms)
-  min: 1000,
+  minTimeout: 1000,
   // max retry delay (ms)
-  max: 1000 * 60,
-  // jitter (%) [0-1]
-  jitter: 0.05,
+  maxTimeout: 1000 * 60,
```
Comment (dlongley) [inline lib/config.js:16]: Since these config names changed, it should be mentioned in the changelog.

File: lib/index.js [hunk 5]
```diff
@@ -3,26 +3,23 @@
...
-
-bedrock.events.on('bedrock.start', () =>
-  bedrock.events.emit('bedrock-redis.ready'));
-
-bedrock.events.on('bedrock.init', () => {
-  api.client = new Client().client;
+export let client = null;
+bedrock.events.on('bedrock.start',
+  () => bedrock.events.emit('bedrock-redis.ready')
+);
```
Comment (dlongley) [inline lib/index.js]:
```suggestion
bedrock.events.on(
  'bedrock.start', () => bedrock.events.emit('bedrock-redis.ready'));
```

File: lib/index.js [hunk 6]
```diff
@@ -3,26 +3,23 @@
...
+bedrock.events.on('bedrock.init', async () => {
+  const clientInstance = new Client();
+  clientInstance.connect();
+  client = clientInstance.client;
```
Comment (dlongley) [inline lib/index.js]: Any reason not to do this?

```suggestion
  client = new Client();
  client.connect();
```

File: lib/index.js [hunk 7]
```diff
@@ -3,26 +3,23 @@
...
 });
+
+export default {
+  client,
+  Client
+};
```
Comment (dlongley) [inline lib/index.js]: No export `default` in new / updated libs. These should be named exports.

File: lib/logger.js [hunk 8]
```diff
@@ -1,6 +1,5 @@
 /*
- * Copyright (c) 2018 Digital Bazaar, Inc. All rights reserved.
+ * Copyright (c) 2012-2022 Digital Bazaar, Inc. All rights reserved.
  */
-const bedrock = require('bedrock');
-
-module.exports = bedrock.loggers.get('app').child('redis');
+import * as bedrock from '@bedrock/core';
+export default bedrock.loggers.get('app').child('redis');
```
Comment (dlongley) [inline lib/logger.js]: No `default` export in new / updated libs, this should be:

```suggestion
export const logger = bedrock.loggers.get('app').child('redis');
```

File: package.json [hunk 9]
```diff
@@ -1,9 +1,14 @@
 {
-  "name": "bedrock-redis",
-  "version": "3.4.2-0",
+  "name": "@bedrock/redis",
+  "version": "4.0.0",
```
Comment (dlongley) [inline package.json]: Don't bump the version manually, it will muck with the publish script.

File: package.json [hunk 10]
```diff
@@ -23,12 +28,18 @@
...
+  },
+  "devDependencies": {
+    "eslint": "^8.18.0",
+    "eslint-config-digitalbazaar": "^4.0.1",
+    "eslint-plugin-jsdoc": "^39.3.3",
+    "eslint-plugin-unicorn": "^43.0.0",
+    "jsdoc": "^3.6.10"
   },
   "directories": {
     "lib": "./lib"
```
Comment (dlongley) [inline package.json:51]: @davidlehn, we also want a `files` entry right?

File: test/test.config.js [hunk 11]
```diff
@@ -0,0 +1,13 @@
+/*
```
Comment (dlongley) [inline test/test.config.js]:
```suggestion
/*!
```

File: test/test.js [hunk 12]
```diff
@@ -0,0 +1,8 @@
+/*
```
Comment (dlongley) [inline test/test.js]:
```suggestion
/*!
```

**Review by dlongley — CHANGES_REQUESTED** [review]

File: lib/client.js (hunk 2 above)
Comment (tolson4) [reply lib/client.js]: Addressed here ee8ede4385fa401ec65af297227714b7043cb771

File: lib/config.js (hunk 4 above)
Comment (tolson4) [reply lib/config.js:16]: ac523cb823c6c92ddaa930dc5aa802587e4bbbff

File: package.json (hunk 9 above)
Comment (tolson4) [reply package.json]: ac523cb823c6c92ddaa930dc5aa802587e4bbbff

File: lib/index.js (hunk 6 above)
Comment (tolson4) [reply lib/index.js]: Yeah, the problem is `client = new Client()` is an instance of the `Client` class implemented by this module, not the node-redis client. I tried to keep the structure that same as it was before, but I'm struggling to see the benefits of including this `Client` class at all. It seems like it would be simpler to just create the node-redis client inline and export that from here instead.

File: lib/client.js (hunk 3 above)
Comment (tolson4) [reply lib/client.js]: resolved here 8dc411e30f97b16e9ab4cd971ee62b1922422e1b

File: lib/index.js (hunk 7 above)
Comment (tolson4) [reply lib/index.js]: 8dc411e30f97b16e9ab4cd971ee62b1922422e1b

File: lib/index.js (hunk 6 above)
Comment (tolson4) [reply lib/index.js]: This has been simplified as part of this commit 8dc411e30f97b16e9ab4cd971ee62b1922422e1b

**Review by dlongley — APPROVED** [review]
LGTM, but I may have missed something -- I chatted with @davidlehn and he said he'd review as well. Once he approves, I'll merge and release this. Thanks!

File: .github/workflows/main.yml [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+      matrix:
+        node-version: [16.x, 18.x]
+    steps:
+    - uses: actions/checkout@v2
+    - name: Use Node.js ${{ matrix.node-version }}
+      uses: actions/setup-node@v1
+      with:
+        node-version: ${{ matrix.node-version }}
+    - run: |
+        npm i --legacy-peer-deps
```
Comment (davidlehn) [inline .github/workflows/main.yml]: In other packages we've been adding a `.npmrc` with `legacy-peer-deps=true`.  That covers both the tests and developers using the flag by default.  (There's probably a better long term solution that we haven't figured out yet.)

File: .gitignore [hunk 14]
```diff
@@ -9,3 +9,4 @@ coverage
 node_modules
 npm-debug.log
 reports
+package-lock.json
```
Comment (davidlehn) [inline .gitignore:12]: Our other packages randomly have this.  Also consider adding `package-lock=false` to your user `~/.npmrc` if you want to always not use lock files by default. Fine to leave it here though.

File: CHANGELOG.md [hunk 15]
```diff
@@ -1,5 +1,14 @@
 # bedrock-redis ChangeLog
 
+## 4.0.0 - 2022-xx-xx
+
+### Changed
+- **BREAKING** Convert to esm.
```
Comment (davidlehn) [inline CHANGELOG.md]: We (I) had been using common text for this. Not a big deal, but could be written as:
```suggestion
- **BREAKING** Convert to module (ESM).
```

File: CHANGELOG.md [hunk 16]
```diff
@@ -1,5 +1,14 @@
...
+- **BREAKING** Update redis client to v4, use node-redis directly.
+- **BREAKING** Use namespaced and up to date bedrock packages throughout.
+- **BREAKING** Remove prefix option, update backoff options.
+- **BREAKING** Change min/max config options to minTimeout/maxTimeout.
+
```
Comment (davidlehn) [inline CHANGELOG.md:22]: Some more changes related to other comments and updates:
```suggestion
- **BREAKING**: Require Node.js >=16.
- Use `package.json` `files` field.

### Added
- Lint code.
- Add tests.

```

File: package.json [hunk 17]
```diff
@@ -1,9 +1,14 @@
 {
-  "name": "bedrock-redis",
+  "name": "@bedrock/redis",
   "version": "3.4.2-0",
   "description": "Bedrock redis module",
   "license": "SEE LICENSE IN LICENSE.md",
-  "main": "./lib",
+  "main": "./lib/index.js",
```
Comment (davidlehn) [inline package.json:6]: Add files field:
```suggestion
  "main": "./lib/index.js",
  "files": [
    "lib/**/*.js",
  ],
```

File: package.json (hunk 10 above)
Comment (davidlehn) [reply package.json:51]: Yes, added suggestion above.

File: package.json [hunk 18]
```diff
@@ -23,12 +28,18 @@
...
-    "bedrock": "1.12.0 - 3.x"
+    "@bedrock/core": "^6.0.1"
+  },
+  "devDependencies": {
+    "eslint": "^8.18.0",
+    "eslint-config-digitalbazaar": "^4.0.1",
+    "eslint-plugin-jsdoc": "^39.3.3",
+    "eslint-plugin-unicorn": "^43.0.0",
+    "jsdoc": "^3.6.10"
   },
```
Comment (davidlehn) [inline package.json:46]: Add engines field:
```suggestion
  },
  "engines": {
    "node": ">=16"
  },
```

File: test/package.json [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+    "c8": "^7.12.0"
+  },
+  "c8": {
+    "excludeNodeModules": false,
+    "include": [
+      "node_modules/@bedrock/redis/**"
+    ],
+    "exclude": [
+      "node_modules/@bedrock/redis/node_modules/**"
+    ]
```
Comment (davidlehn) [inline test/package.json:31]: Matching pattern used in other modules:
```suggestion
    ],
    "reporter": [
      "lcov",
      "text-summary",
      "text"
    ]
```

File: test/package.json [hunk 20]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "@bedrock/redis-test",
+  "version": "1.0.0",
+  "description": "Test suites for bedrock-redis",
+  "type": "module",
+  "scripts": {
+    "test": "node --preserve-symlinks test.js test",
+    "coverage": "c8 --reporter=text --reporter=text-summary npm test",
+    "coverage-ci": "c8 --reporter=lcov npm test",
```
Comment (davidlehn) [inline test/package.json]: Matching pattern used in other modules:
```suggestion
    "coverage": "c8 npm test",
    "coverage-ci": "c8 --reporter=lcovonly --reporter=text --reporter=text-summary npm test",
```

File: test/package.json [hunk 21]
```diff
@@ -0,0 +1,27 @@
+{
+  "name": "@bedrock/redis-test",
+  "version": "1.0.0",
+  "description": "Test suites for bedrock-redis",
+  "type": "module",
```
Comment (davidlehn) [inline test/package.json:5]:
```suggestion
  "type": "module",
  "private": true,
```

File: .github/workflows/main.yml [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+    services:
+      redis:
+        image: redis:6.0
+        ports:
+          - 6379:6379
+    strategy:
+      matrix:
+        node-version: [16.x, 18.x]
```
Comment (davidlehn) [inline .github/workflows/main.yml]: Only need to run coverage once:
```suggestion
        node-version: [18.x]
```

File: lib/client.js [hunk 23]
```diff
@@ -1,68 +1,34 @@
...
-      jitter: config.redis.retry.jitter,
-      factor: config.redis.retry.factor
-    });
-    logger.verbose(`initializing client: ${host}:${port}`);
-    this.client = redis.createClient({
+export async function createClient({
+  host = config.redis.host,
+  port = config.redis.port,
+} = {}) {
+  logger.debug(`Initializing redis client at: ${host}:${port}`);
```
Comment (davidlehn) [inline lib/client.js]: `redis` in the string not needed here since the child logger is named `redis`.

File: lib/config.js [hunk 24]
```diff
@@ -1,33 +1,21 @@
...
 // retry backoff control
-config.redis.retry = {
+config.redis.retryOptions = {
   // min retry delay (ms)
-  min: 1000,
+  minTimeout: 1000,
   // max retry delay (ms)
-  max: 1000 * 60,
-  // jitter (%) [0-1]
-  jitter: 0.05,
```
Comment (davidlehn) [inline lib/config.js:26]: Looks like the `retry` package has a `randomize` option.  Not as flexible as the `jitter` one, but maybe should be set or allowed to be set?  Idea was to avoid dogpiling effects if many clients try to connect at the same time.

File: CHANGELOG.md [hunk 25]
```diff
@@ -1,5 +1,14 @@
 # bedrock-redis ChangeLog
 
+## 4.0.0 - 2022-xx-xx
+
+### Changed
```
Comment (davidlehn) [inline CHANGELOG.md:5]:
```suggestion
### Changed
- **BREAKING**: Rename package from `bedrock-redis` to `@bedrock/redis`.
```

File: CHANGELOG.md [hunk 26]
```diff
@@ -1,5 +1,14 @@
...
+- **BREAKING** Convert to esm.
+- **BREAKING** Update redis client to v4, use node-redis directly.
+- **BREAKING** Use namespaced and up to date bedrock packages throughout.
+- **BREAKING** Remove prefix option, update backoff options.
+- **BREAKING** Change min/max config options to minTimeout/maxTimeout.
```
Comment (davidlehn) [inline CHANGELOG.md]: Be more explicit with what needs updating:
```suggestion
- **BREAKING** Remove `prefix` option.
- **BREAKING** Update retry options due to changing from `backo` package to `retry`:
  - Rename `bedrock.config.redis.retry` to `bedrock.config.redis.retryOptions`.
  - Rename `min`/`max` to `minTimeout`/`maxTimeout`.
  - Rename `maxTimesConnected` to `retries`.
  - Remove `jitter` and `maxTotal`.
```

File: lib/client.js [hunk 27]
```diff
@@ -1,68 +1,34 @@
...
-          logger.error('exceeded max connection times', {
-            totalRetryTime: options.total_retry_time,
-            timesConnected: options.times_connected
-          });
-          // FIXME: improve error handling
-          return new Error('redis: exceed max connection times');
+      reconnectStrategy: attempt => {
+        const max = config.redis.retryOptions.retries;
+        if(attempt > max) {
+          return new Error(`Exceeded max (${max}) connection attempts.`);
```
Comment (davidlehn) [inline lib/client.js:21]: I'm not sure how this propagates up, would it be useful to preserve the previous logging message?

File: lib/client.js [hunk 28]
```diff
@@ -1,68 +1,34 @@
...
         }
-        // reconnect delay with backoff
-        if(options.attempt === 1) {
-          backoff.reset();
-        }
-        const delay = backoff.duration();
-        logger.info(`will reconnect in ${delay}ms`);
```
Comment (davidlehn) [inline lib/client.js:58]: Should this stay around?  (perhaps with attempt number vs delay)

File: lib/config.js [hunk 29]
```diff
@@ -1,33 +1,21 @@
...
-// A string used to prefix all used keys (e.g. namespace:test).
-// `prefix` is useful when multiple bedrock applications share the same redis
-// instance in order to prevent key collisions and simplify debugging.
-// `prefix` should be as few characters as possible because the memory
-// footprint of every key will be increased by one byte per character.
-config.redis.prefix = null;
-
 // retry backoff control
-config.redis.retry = {
+config.redis.retryOptions = {
```
Comment (davidlehn) [inline lib/config.js]: Why was this renamed?  From one viewpoint, everything in the config is an option.

**Review by davidlehn — COMMENTED** [review]
That retry logic is difficult to review with both API and retry lib changes at once.  I recall it was tricky to get right, but it was addressing ability to debug connection problems from long ago.  Does seem maybe more logging is appropriate since if you have connection problems you may need to know it's trying.

File: test/package.json (hunk 19 above)
Comment (tolson4) [reply test/package.json:31]: Why put the reporters here and in the scripts? Which takes precedent, the ones from the config or the ones specified in the commands?

Comment (davidlehn) [reply test/package.json:31]: Using `--reporter` will override these defaults.  The theory here was to add defaults that would work for the common "coverage" and "coverage-report" scripts, and then add all `--reporter` options to "coverage-ci" but use `lcovonly`. 
 `lcovonly` omits html output, which is of no use for the CI target.  But I've found it's useful to see the text output in CI output, so had to add all three reporters there.  Having defaults seemed useful for the (rare) times when you might run c8/nyc by hand with some other option but wanted the default reporters.

File: lib/config.js (hunk 24 above)
Comment (tolson4) [reply lib/config.js:26]: Resolved here 64e5a2ce36c8aeff4ebae2e29f2a46dde74e55f2

File: lib/config.js (hunk 29 above)
Comment (tolson4) [reply lib/config.js]: Resolved here 64e5a2ce36c8aeff4ebae2e29f2a46dde74e55f2

The reason it was renamed was because `retry.retries` sounded funky to me. But `retryOptions` just moved the problem to a different place as you mentioned.

File: lib/client.js (hunk 23 above)
Comment (tolson4) [reply lib/client.js]: resolved in 7433ea68e47d9a51d63e6ae851b07f141a62db45

File: lib/client.js (hunk 28 above)
Comment (tolson4) [reply lib/client.js:58]: resolved in 7433ea68e47d9a51d63e6ae851b07f141a62db45

File: lib/client.js (hunk 27 above)
Comment (tolson4) [reply lib/client.js:21]: This propagates up to the `_init()` function in `index.js` which does log the error.

```js
async function _init() {
  try {
    _client = await createClient();
  } catch(error) {
    logger.error('Could not connect to redis', {error});
    throw new BedrockError(
      'Could not connect to redis.', {
        host: config.redis.host,
        port: config.redis.port
      }, error);
  }
}
```

In the tests this is caught by the test case and the error is matched to ensure the correct number of retries took place
https://github.com/dogwoodlogic/bedrock-redis/actions/runs/3174561898/jobs/5171513587#step:6:18
https://github.com/dogwoodlogic/bedrock-redis/blob/updates/test/mocha/10-api.js#L43

File: .gitignore (hunk 14 above)
Comment (tolson4) [reply .gitignore:12]: There are some other things I develop on that do use `package-lock`. I have been adding `package-lock=false` locally to a projects npmrc when the lockfile is not needed. DB might consider something like this since the default behavior of NPM is to include the package-lock. I would assume most outside devs will not have changed this setting in their root npmrc, so it may be worthwhile to add directly into each projects npmrc.

File: .github/workflows/main.yml (hunk 13 above)
Comment (tolson4) [reply .github/workflows/main.yml]: resolved here deb99e8666fa0b81164944ceca67fc4550f24ff6

**Review by davidlehn — APPROVED** [review]
